File: packages/es-markdown-core/src/tree-dumper.ts
import type { Attributes, Dimensions, Metadata, MetadataValue } from "./values.js";
import type { Citation } from "./model/cite.js";
import type { Footnote } from "./model/footnote.js";
import type { Markup } from "./model/markup.js";
import type { Specimen } from "./model/specimen.js";
import type { CitationReferent, Destination, OrderedListDelimiter, OrderedListVariant, Scope } from "./values.js";
import { visit, type Visitor } from "./visitor.js";

/** Produces the canonical debug tree for immutable Markdown markup. */
export class TreeDumper {
    private constructor() {}

    /** Returns the canonical debug dump for `root` and its owned markup. */
    static dump(root: Markup): string {
        const state = new State();
        state.dump(root);
        return state.result();
    }
}

class State {
    private readonly remainingNodes: number[] = [];
    private readonly lines: string[] = [];

    /** Each callback emits exactly its node and chooses its children/fields. */
    private readonly visitor: Visitor<void> = {
        document: (node) => {
            // The footnotes are value lines after the content, never counted
            // by the document's own `children`.
            this.line("Document", node, [], node.content.length);
            this.nested(
                node.content.length + node.footnotes.length + node.specimens.length + (node.metadata === null ? 0 : 1),
                () => {
                    if (node.metadata !== null) this.metadata(node.metadata);
                    for (const child of node.content) this.dump(child);
                    for (const footnote of node.footnotes) this.footnote(footnote);
                    for (const specimen of node.specimens) this.specimen(specimen);
                }
            );
        },
        callout: (node) => {
            this.line(
                "Callout",
                node,
                [`variant=${optional(node.variant)}`, `collapsed=${node.collapsed ?? "null"}`],
                node.content.length
            );
            // A non-null title is a `Title` group before the content; a null
            // one prints nothing. Neither is counted by `children`.
            this.nested(node.content.length + (node.title === null ? 0 : 1), () => {
                if (node.title !== null) {
                    this.group("Title", node.title.length);
                    this.nested(node.title.length, () => {
                        for (const child of node.title ?? []) this.dump(child);
                    });
                }
                for (const child of node.content) this.dump(child);
            });
        },
        paragraph: (node) => this.container("Paragraph", node, [], node.content),
        heading: (node) => this.container("Heading", node, [`level=${node.level}`], node.content),
        thematicBreak: (node) => this.line("ThematicBreak", node),
        list: (node) =>
            this.container(
                "List",
                node,
                [
                    `flavor=${node.flavor}`,
                    `start=${node.start ?? "null"}`,
                    `variant=${variant(node.variant)}`,
                    `delimiter=${delimiter(node.delimiter)}`,
                    `tight=${node.tight}`
                ],
                node.items
            ),
        listItem: (node) => this.container("ListItem", node, [`marker=${optional(node.marker)}`], node.content),
        codeBlock: (node) =>
            this.line("CodeBlock", node, [
                `info=${optional(node.info)}`,
                `language=${optional(node.language)}`,
                `literal=${escaped(node.literal)}`,
                `fenced=${node.fenced}`,
                `closed=${node.closed}`
            ]),
        htmlBlock: (node) => this.line("HTMLBlock", node, [`literal=${escaped(node.literal)}`]),
        formulaBlock: (node) => this.line("FormulaBlock", node, [`literal=${escaped(node.literal)}`]),
        table: (node) => {
            const columns = node.columns.map((column) => `${column.flow}:${column.relative ?? "null"}`).join(",");
            this.line(
                "Table",
                node,
                [`columns=[${columns}]`],
                node.head.length + node.content.length + node.foot.length
            );
            this.nested(3 + (node.caption ? 1 : 0), () => {
                if (node.caption) this.dump(node.caption);
                for (const [name, rows] of [
                    ["TableHead", node.head],
                    ["TableBody", node.content],
                    ["TableFoot", node.foot]
                ] as const) {
                    this.group(name, rows.length);
                    this.nested(rows.length, () => {
                        for (const row of rows) this.dump(row);
                    });
                }
            });
        },
        tableCaption: (node) => this.container("TableCaption", node, [], node.content),
        tableRow: (node) => this.container("TableRow", node, [], node.cells),
        tableCell: (node) =>
            this.container("TableCell", node, [`rowspan=${node.rowspan}`, `colspan=${node.colspan}`], node.content),
        definitionList: (node) => this.container("DefinitionList", node, [], node.definitions),
        definition: (node) => {
            this.line("Definition", node, [`compact=${node.compact}`], node.content.length);
            this.nested(node.content.length + 1, () => {
                this.group("DefinitionTerm", node.term.length);
                this.nested(node.term.length, () => node.term.forEach((child) => this.dump(child)));
                for (const body of node.content) {
                    this.group("DefinitionBody", body.length);
                    this.nested(body.length, () => body.forEach((child) => this.dump(child)));
                }
            });
        },
        directiveBlock: (node) => {
            this.line("DirectiveBlock", node, [`name=${optional(node.name)}`], node.content.length);
            this.nested(node.content.length + (node.label === null ? 0 : 1), () => {
                if (node.label !== null) this.dump(node.label);
                for (const child of node.content) this.dump(child);
            });
        },
        directiveLabel: (node) => this.container("DirectiveLabel", node, [], node.content),
        text: (node) => this.line("Text", node, [`literal=${escaped(node.literal)}`]),
        softBreak: (node) => this.line("SoftBreak", node),
        lineBreak: (node) => this.line("LineBreak", node),
        code: (node) => this.line("Code", node, [`literal=${escaped(node.literal)}`]),
        html: (node) => this.line("HTML", node, [`literal=${escaped(node.literal)}`]),
        crossLink: (node) =>
            this.line("CrossLink", node, [`dest=${destination(node.dest)}`, `label=${optional(node.label)}`]),
        crossEmbedded: (node) =>
            this.line("CrossEmbedded", node, [
                `dest=${destination(node.dest)}`,
                `label=${optional(node.label)}`,
                `dimensions=${dimensions(node.dimensions)}`
            ]),
        comment: (node) => this.line("Comment", node, [`literal=${escaped(node.literal)}`]),
        formula: (node) => this.line("Formula", node, [`mode=${node.mode}`, `literal=${escaped(node.literal)}`]),
        emphasis: (node) => this.container("Emphasis", node, [], node.content),
        strong: (node) => this.container("Strong", node, [], node.content),
        strikethrough: (node) => this.container("Strikethrough", node, [], node.content),
        mark: (node) => this.container("Mark", node, [], node.content),
        insertion: (node) => this.container("Insertion", node, [], node.content),
        span: (node) => this.container("Span", node, [], node.content),
        superscript: (node) => this.container("Superscript", node, [], node.content),
        subscript: (node) => this.container("Subscript", node, [], node.content),
        link: (node) =>
            this.container(
                "Link",
                node,
                [`dest=${destination(node.dest)}`, `title=${optional(node.title)}`],
                node.content
            ),
        embedded: (node) =>
            this.container(
                "Embedded",
                node,
                [
                    `dest=${destination(node.dest)}`,
                    `title=${optional(node.title)}`,
                    `dimensions=${dimensions(node.dimensions)}`
                ],
                node.content
            ),
        directive: (node) => {
            this.line("Directive", node, [`name=${escaped(node.name)}`]);
            this.nested(node.label === null ? 0 : 1, () => {
                if (node.label !== null) this.dump(node.label);
            });
        },
        cite: (node) => {
            // Each item is a value line whose affixes are groups; the cite's
            // own `children` counts the items.
            this.line("Cite", node, [], node.citations.length);
            this.nested(node.citations.length, () => {
                for (const item of node.citations) this.citation(item);
            });
        }
    };

    dump(node: Markup): void {
        visit(node, this.visitor);
    }

    result(): string {
        return `${this.lines.join("\n")}\n`;
    }

    private citation(item: Citation): void {
        this.value("Citation", item.scope, [`referent=${referent(item.referent)}`], 0);
        this.nested(2, () => {
            this.group("CitationPrefix", item.prefix.length);
            this.nested(item.prefix.length, () => {
                for (const child of item.prefix) this.dump(child);
            });
            this.group("CitationSuffix", item.suffix.length);
            this.nested(item.suffix.length, () => {
                for (const child of item.suffix) this.dump(child);
            });
        });
    }

    private metadata(value: Metadata): void {
        this.value(
            "Metadata",
            value.scope,
            [
                `name=${value.name === null ? "null" : metadataValue(value.name)}`,
                `title=${value.title === null ? "null" : metadataValue(value.title)}`,
                `subtitle=${value.subtitle === null ? "null" : metadataValue(value.subtitle)}`,
                `time=${value.time === null ? "null" : metadataValue(value.time)}`,
                `date=${value.date === null ? "null" : metadataValue(value.date)}`,
                `authors=${value.authors === null ? "null" : metadataValue(value.authors)}`,
                `keywords=${value.keywords === null ? "null" : metadataValue(value.keywords)}`,
                `abstract=${value.abstract === null ? "null" : metadataValue(value.abstract)}`,
                `state=${value.state === null ? "null" : metadataValue(value.state)}`,
                `comment=${value.comment === null ? "null" : metadataValue(value.comment)}`
            ],
            0
        );
    }

    private footnote(value: Footnote): void {
        this.value("Footnote", value.scope, [`id=${escaped(value.id)}`], value.content.length);
        this.nested(value.content.length, () => {
            for (const child of value.content) this.dump(child);
        });
    }

    private specimen(value: Specimen): void {
        this.value(
            "Specimen",
            value.scope,
            [`id=${value.id === null ? "null" : escaped(value.id)}`, `start=${value.start ?? "null"}`],
            value.content.length
        );
        this.nested(value.content.length, () => {
            for (const child of value.content) this.dump(child);
        });
    }

    private container(kind: string, node: Markup, fields: readonly string[], children: readonly Markup[]): void {
        this.line(kind, node, fields, children.length);
        this.nested(children.length, () => {
            for (const child of children) this.dump(child);
        });
    }

    private line(kind: string, node: Markup, fields: readonly string[] = [], children = 0): void {
        this.value(
            kind,
            node.scope,
            [`anchor=${optional(node.anchor)}`, `attributes=${attributes(node.attributes)}`, ...fields],
            children
        );
    }

    /** A value line prints like a node line: scope, fields, `children`. */
    private value(kind: string, at: Scope, fields: readonly string[], children: number): void {
        const fieldText = fields.length === 0 ? "" : ` ${fields.join(" ")}`;
        this.emit(`${kind} ${scope(at)}${fieldText} children=${children}`);
    }

    /**
     * A group line nests a node-valued list under its owner: `Kind children=N`
     * with no scope and no fields. The caller opens the list's own nesting.
     */
    private group(kind: string, children: number): void {
        this.emit(`${kind} children=${children}`);
    }

    private emit(text: string): void {
        if (this.remainingNodes.length === 0) {
            this.lines.push(text);
            return;
        }

        const parent = this.remainingNodes.length - 1;
        const prefix = this.remainingNodes
            .slice(0, -1)
            .map((remaining) => (remaining > 0 ? "│   " : "    "))
            .join("");
        const connector = this.remainingNodes[parent] === 1 ? "└── " : "├── ";
        this.lines.push(prefix + connector + text);
        this.remainingNodes[parent] = this.remainingNodes[parent]! - 1;
    }

    private nested(count: number, body: () => void): void {
        this.remainingNodes.push(count);
        body();
        if (this.remainingNodes.pop() !== 0) throw new Error("node dumper did not emit every owned node");
    }
}

function scope(value: Scope): string {
    return `scope=${value.start.line}:${value.start.column}..${value.end.line}:${value.end.column}`;
}

function optional(value: string | null): string {
    return value === null ? "null" : escaped(value);
}

function delimiter(value: OrderedListDelimiter | null): string {
    if (value === null || typeof value === "string") return value ?? "null";
    return `parenthesis(closed=${value.closed})`;
}

function variant(value: OrderedListVariant | null): string {
    if (value === null || typeof value === "string") return value ?? "null";
    return `${value.kind}(lowercased=${value.lowercased})`;
}

/** A tagged value prints its branch and its named fields with no spaces. */
function referent(value: CitationReferent): string {
    return value.kind === "bib"
        ? `bib(key=${escaped(value.key)},mode=${value.mode})`
        : `${value.kind}(id=${escaped(value.id)})`;
}

/** A tagged value prints its branch and its named fields with no spaces. */
function destination(value: Destination): string {
    return value.kind === "url"
        ? `url(${escaped(value.value)})`
        : `cross(path=${escaped(value.path)},anchor=${optional(value.anchor)})`;
}

function escaped(value: string): string {
    return JSON.stringify(value);
}

function attributes(value: Attributes): string {
    return (
        "{" +
        [
            ...value.classes.map(
                (name) => "." + (/^[!-~]+$/u.test(name) && !/["\\{}[\]()=]/u.test(name) ? name : escaped(name))
            ),
            ...value.records.map((record) => `${record.name}=${escaped(record.value)}`)
        ].join(" ") +
        "}"
    );
}
function metadataValue(value: MetadataValue): string {
    if (value.kind === "list")
        return "list([" + value.items.map((item) => `${item.kind}(${escaped(item.value)})`).join(",") + "])";
    const scalar = value.value;
    if (scalar.kind === "null") return "scalar(null)";
    return `scalar(${scalar.kind}(${scalar.kind === "bool" ? String(scalar.value) : escaped(scalar.value)}))`;
}

function dimensions(value: Dimensions | null): string {
    return value === null ? "null" : `(width=${value.width},height=${value.height ?? "null"})`;
}

File: packages/kotlin-markdown-core/src/commonMain/kotlin/com/nouprax/markdown/core/visitor/TreeDumper.kt
package com.nouprax.markdown.core

/** Produces the canonical debug tree for immutable Markdown markup. */
public object TreeDumper {
    /** Returns the canonical debug dump for [root] and its owned markup. */
    public fun dump(root: Markup): String =
        State().run {
            dump(root)
            result()
        }
}

private class State {
    private val remainingNodes = mutableListOf<Int>()
    private val lines = mutableListOf<String>()
    private val visitor = DumpVisitor(this)

    fun dump(node: Markup) {
        node.accept(visitor)
    }

    fun result(): String = lines.joinToString(separator = "\n", postfix = "\n")

    fun container(
        kind: String,
        node: Markup,
        fields: kotlin.collections.List<String> = emptyList(),
        children: kotlin.collections.List<Markup>,
    ) {
        line(kind, node, fields, children.size)
        nested(children.size) { children.forEach(::dump) }
    }

    fun line(
        kind: String,
        node: Markup,
        fields: kotlin.collections.List<String> = emptyList(),
        children: Int = 0,
    ) {
        line(
            kind,
            node.scope,
            listOf("anchor=${optional(node.anchor)}", "attributes=${attributes(node.attributes)}") + fields,
            children,
        )
    }

    /** A value line has the node line's shape: a scoped value prints like a node. */
    fun line(
        kind: String,
        scope: Scope,
        fields: kotlin.collections.List<String>,
        children: Int,
    ) {
        val fieldText = if (fields.isEmpty()) "" else " ${fields.joinToString(" ")}"
        emit("$kind ${scope(scope)}$fieldText children=$children")
    }

    /**
     * A group line nests a node-valued list under its owner: `Kind children=N`
     * with no scope and no fields. The caller opens the list's own nesting.
     */
    fun group(
        kind: String,
        children: Int,
    ) {
        emit("$kind children=$children")
    }

    private fun emit(text: String) {
        if (remainingNodes.isEmpty()) {
            lines += text
            return
        }

        val parent = remainingNodes.lastIndex
        val prefix = remainingNodes.dropLast(1).joinToString("") { if (it > 0) "│   " else "    " }
        val connector = if (remainingNodes[parent] == 1) "└── " else "├── "
        lines += prefix + connector + text
        remainingNodes[parent] -= 1
    }

    fun nested(
        count: Int,
        body: () -> Unit,
    ) {
        remainingNodes += count
        body()
        check(remainingNodes.removeAt(remainingNodes.lastIndex) == 0) {
            "node dumper did not emit every owned node"
        }
    }
}

/** Each visit emits exactly that node and chooses its children and fields. */
private class DumpVisitor(
    private val state: State,
) : Visitor<Unit> {
    override fun visit(document: Document) {
        // The footnotes are value lines after the content, each nesting its
        // own content; `children` counts the content alone.
        state.line("Document", document, children = document.content.size)
        state.nested(
            document.content.size + document.footnotes.size + document.specimens.size +
                (if (document.metadata == null) 0 else 1),
        ) {
            document.metadata?.let { metadata(it) }
            document.content.forEach(state::dump)
            document.footnotes.forEach { footnote(it) }
            document.specimens.forEach { specimen(it) }
        }
    }

    private fun metadata(value: Metadata) {
        state.line(
            "Metadata",
            value.scope,
            listOf(
                "name=${value.name?.let(::metadataValue) ?: "null"}",
                "title=${value.title?.let(::metadataValue) ?: "null"}",
                "subtitle=${value.subtitle?.let(::metadataValue) ?: "null"}",
                "time=${value.time?.let(::metadataValue) ?: "null"}",
                "date=${value.date?.let(::metadataValue) ?: "null"}",
                "authors=${value.authors?.let(::metadataValue) ?: "null"}",
                "keywords=${value.keywords?.let(::metadataValue) ?: "null"}",
                "abstract=${value.`abstract`?.let(::metadataValue) ?: "null"}",
                "state=${value.state?.let(::metadataValue) ?: "null"}",
                "comment=${value.comment?.let(::metadataValue) ?: "null"}",
            ),
            0,
        )
    }

    private fun footnote(value: Footnote) {
        state.line("Footnote", value.scope, listOf("id=${escaped(value.id)}"), value.content.size)
        state.nested(value.content.size) { value.content.forEach(state::dump) }
    }

    private fun specimen(value: Specimen) {
        state.line(
            "Specimen",
            value.scope,
            listOf("id=${optional(value.id)}", "start=${value.start ?: "null"}"),
            value.content.size,
        )
        state.nested(value.content.size) { value.content.forEach(state::dump) }
    }

    override fun visit(callout: Callout) {
        state.line(
            "Callout",
            callout,
            listOf("variant=${optional(callout.variant)}", "collapsed=${callout.collapsed ?: "null"}"),
            callout.content.size,
        )
        // A non-null title is a `Title` group before the content; a
        // null one prints nothing. Neither is counted by `children`.
        state.nested(callout.content.size + (if (callout.title == null) 0 else 1)) {
            callout.title?.let { title ->
                state.group("Title", title.size)
                state.nested(title.size) { title.forEach(state::dump) }
            }
            callout.content.forEach(state::dump)
        }
    }

    override fun visit(paragraph: Paragraph) {
        state.container("Paragraph", paragraph, children = paragraph.content)
    }

    override fun visit(heading: Heading) {
        state.container("Heading", heading, listOf("level=${heading.level}"), heading.content)
    }

    override fun visit(thematicBreak: ThematicBreak) {
        state.line("ThematicBreak", thematicBreak)
    }

    override fun visit(list: List) {
        state.container(
            "List",
            list,
            listOf(
                "flavor=${list.flavor.token()}",
                "start=${list.start ?: "null"}",
                "variant=${list.variant?.token() ?: "null"}",
                "delimiter=${list.delimiter?.token() ?: "null"}",
                "tight=${list.tight}",
            ),
            list.items,
        )
    }

    override fun visit(listItem: ListItem) {
        state.container(
            "ListItem",
            listItem,
            listOf("marker=${optional(listItem.marker)}"),
            listItem.content,
        )
    }

    override fun visit(codeBlock: CodeBlock) {
        state.line(
            "CodeBlock",
            codeBlock,
            listOf(
                "info=${optional(codeBlock.info)}",
                "language=${optional(codeBlock.language)}",
                "literal=${escaped(codeBlock.literal)}",
                "fenced=${codeBlock.fenced}",
                "closed=${codeBlock.closed}",
            ),
        )
    }

    override fun visit(htmlBlock: HTMLBlock) {
        state.line("HTMLBlock", htmlBlock, listOf("literal=${escaped(htmlBlock.literal)}"))
    }

    override fun visit(formulaBlock: FormulaBlock) {
        state.line("FormulaBlock", formulaBlock, listOf("literal=${escaped(formulaBlock.literal)}"))
    }

    override fun visit(table: Table) {
        val columns =
            table.columns.joinToString(
                ",",
            ) { "${it.flow.token()}:${it.relative?.let(::decimal) ?: "null"}" }
        state.line("Table", table, listOf("columns=[$columns]"), table.head.size + table.content.size + table.foot.size)
        state.nested(3 + if (table.caption == null) 0 else 1) {
            table.caption?.let(state::dump)
            for ((name, rows) in listOf(
                "TableHead" to table.head,
                "TableBody" to table.content,
                "TableFoot" to table.foot,
            )) {
                state.group(name, rows.size)
                state.nested(rows.size) { rows.forEach(state::dump) }
            }
        }
    }

    override fun visit(tableCaption: TableCaption): Unit =
        state.container("TableCaption", tableCaption, emptyList(), tableCaption.content)

    override fun visit(tableRow: TableRow): Unit = state.container("TableRow", tableRow, emptyList(), tableRow.cells)

    override fun visit(tableCell: TableCell): Unit =
        state.container(
            "TableCell",
            tableCell,
            listOf("rowspan=${tableCell.rowspan}", "colspan=${tableCell.colspan}"),
            tableCell.content,
        )

    override fun visit(directiveBlock: DirectiveBlock) {
        state.line(
            "DirectiveBlock",
            directiveBlock,
            listOf("name=${optional(directiveBlock.name)}"),
            children = directiveBlock.content.size,
        )
        state.nested(directiveBlock.content.size + if (directiveBlock.label == null) 0 else 1) {
            directiveBlock.label?.let(state::dump)
            directiveBlock.content.forEach(state::dump)
        }
    }

    override fun visit(directiveLabel: DirectiveLabel) {
        state.container("DirectiveLabel", directiveLabel, children = directiveLabel.content)
    }

    override fun visit(text: Text) {
        state.line("Text", text, listOf("literal=${escaped(text.literal)}"))
    }

    override fun visit(softBreak: SoftBreak) {
        state.line("SoftBreak", softBreak)
    }

    override fun visit(lineBreak: LineBreak) {
        state.line("LineBreak", lineBreak)
    }

    override fun visit(code: Code) {
        state.line("Code", code, listOf("literal=${escaped(code.literal)}"))
    }

    override fun visit(html: HTML) {
        state.line("HTML", html, listOf("literal=${escaped(html.literal)}"))
    }

    override fun visit(crossLink: CrossLink) {
        state.line(
            "CrossLink",
            crossLink,
            listOf("dest=${destination(crossLink.dest)}", "label=${optional(crossLink.label)}"),
        )
    }

    override fun visit(crossEmbedded: CrossEmbedded) {
        state.line(
            "CrossEmbedded",
            crossEmbedded,
            listOf(
                "dest=${destination(crossEmbedded.dest)}",
                "label=${optional(crossEmbedded.label)}",
                "dimensions=${dimensions(crossEmbedded.dimensions)}",
            ),
        )
    }

    override fun visit(comment: Comment) {
        state.line("Comment", comment, listOf("literal=${escaped(comment.literal)}"))
    }

    override fun visit(formula: Formula) {
        state.line("Formula", formula, listOf("mode=${formula.mode.token()}", "literal=${escaped(formula.literal)}"))
    }

    override fun visit(emphasis: Emphasis) {
        state.container("Emphasis", emphasis, children = emphasis.content)
    }

    override fun visit(strong: Strong) {
        state.container("Strong", strong, children = strong.content)
    }

    override fun visit(strikethrough: Strikethrough) {
        state.container("Strikethrough", strikethrough, children = strikethrough.content)
    }

    override fun visit(mark: Mark) {
        state.container("Mark", mark, children = mark.content)
    }

    override fun visit(insertion: Insertion) {
        state.container("Insertion", insertion, children = insertion.content)
    }

    override fun visit(span: Span) {
        state.container("Span", span, children = span.content)
    }

    override fun visit(superscript: Superscript) {
        state.container("Superscript", superscript, children = superscript.content)
    }

    override fun visit(definitionList: DefinitionList) {
        state.container("DefinitionList", definitionList, children = definitionList.definitions)
    }

    override fun visit(definition: Definition) {
        state.line("Definition", definition, listOf("compact=${definition.compact}"), definition.content.size)
        state.nested(definition.content.size + 1) {
            state.group("DefinitionTerm", definition.term.size)
            state.nested(definition.term.size) { definition.term.forEach(state::dump) }
            for (body in definition.content) {
                state.group("DefinitionBody", body.size)
                state.nested(body.size) { body.forEach(state::dump) }
            }
        }
    }

    override fun visit(subscript: Subscript) {
        state.container("Subscript", subscript, children = subscript.content)
    }

    override fun visit(link: Link) {
        state.container(
            "Link",
            link,
            listOf(
                "dest=${destination(link.dest)}",
                "title=${optional(link.title)}",
            ),
            link.content,
        )
    }

    override fun visit(embedded: Embedded) {
        state.container(
            "Embedded",
            embedded,
            listOf(
                "dest=${destination(embedded.dest)}",
                "title=${optional(embedded.title)}",
                "dimensions=${dimensions(embedded.dimensions)}",
            ),
            embedded.content,
        )
    }

    override fun visit(directive: Directive) {
        state.line("Directive", directive, listOf("name=${escaped(directive.name)}"))
        state.nested(if (directive.label == null) 0 else 1) {
            directive.label?.let(state::dump)
        }
    }

    override fun visit(cite: Cite) {
        // The items are value lines under the cite, and `children` counts
        // them; each item's affixes are groups whose nodes nest below them.
        state.line("Cite", cite, children = cite.citations.size)
        state.nested(cite.citations.size) { cite.citations.forEach { citation(it) } }
    }

    private fun citation(value: Citation) {
        state.line("Citation", value.scope, listOf("referent=${referent(value.referent)}"), 0)
        state.nested(2) {
            state.group("CitationPrefix", value.prefix.size)
            state.nested(value.prefix.size) { value.prefix.forEach(state::dump) }
            state.group("CitationSuffix", value.suffix.size)
            state.nested(value.suffix.size) { value.suffix.forEach(state::dump) }
        }
    }
}

private fun scope(value: Scope): String =
    "scope=${value.start.line}:${value.start.column}..${value.end.line}:${value.end.column}"

private fun optional(value: String?): String = value?.let(::escaped) ?: "null"

/** A tagged value prints as its branch name applied to its fields, in declaration order. */
private fun destination(value: Destination): String =
    when (value) {
        is Destination.Url -> "url(${escaped(value.value)})"
        is Destination.Cross -> "cross(path=${escaped(value.path)},anchor=${optional(value.anchor)})"
    }

private fun referent(value: CitationReferent): String =
    when (value) {
        is CitationReferent.Bib -> "bib(key=${escaped(value.key)},mode=${value.mode.token()})"
        is CitationReferent.Footnote -> "footnote(id=${escaped(value.id)})"
        is CitationReferent.Specimen -> "specimen(id=${escaped(value.id)})"
    }

private fun BibMode.token(): String =
    when (this) {
        BibMode.NORMAL -> "normal"
        BibMode.AUTHOR_IN_TEXT -> "authorInText"
        BibMode.SUPPRESS_AUTHOR -> "suppressAuthor"
    }

private fun Placement.token(): String = name.lowercase()

private fun ListFlavor.token(): String = name.lowercase()

private fun OrderedListVariant.token(): String =
    when (this) {
        OrderedListVariant.Decimal -> "decimal"
        is OrderedListVariant.Alpha -> "alpha(lowercased=$lowercased)"
        is OrderedListVariant.Roman -> "roman(lowercased=$lowercased)"
        OrderedListVariant.Default -> "default"
    }

private fun OrderedListDelimiter.token(): String =
    when (this) {
        OrderedListDelimiter.Period -> "period"
        is OrderedListDelimiter.Parenthesis -> "parenthesis(closed=$closed)"
        OrderedListDelimiter.Default -> "default"
    }

private fun Flow.token(): String = name.lowercase()

private fun escaped(value: String): String =
    buildString {
        append('"')
        value.forEach { character ->
            when (character) {
                '"' -> {
                    append("\\\"")
                }

                '\\' -> {
                    append("\\\\")
                }

                '\b' -> {
                    append("\\b")
                }

                '\u000c' -> {
                    append("\\f")
                }

                '\n' -> {
                    append("\\n")
                }

                '\r' -> {
                    append("\\r")
                }

                '\t' -> {
                    append("\\t")
                }

                else -> {
                    if (character.code < 0x20) {
                        append("\\u")
                        append(character.code.toString(16).padStart(4, '0'))
                    } else {
                        append(character)
                    }
                }
            }
        }
        append('"')
    }

/** Emit shortest round-trip digits with the same notation thresholds on every runtime. */
private fun decimal(value: Double): String {
    val parts = value.toString().lowercase().split('e')
    val mantissa = parts[0].split('.')
    var digits = mantissa.joinToString("")
    var point = mantissa[0].length + if (parts.size == 2) parts[1].toInt() else 0
    while (digits.startsWith('0')) {
        digits = digits.drop(1)
        point--
    }
    digits = digits.trimEnd('0')
    // Some runtimes print two significant digits for subnormal values where
    // one already round-trips. Test rounded prefixes, preserving the value.
    for (length in 1 until digits.length) {
        val rounded = digits.take(length).toLong() + if (digits[length] >= '5') 1 else 0
        val candidate = rounded.toString()
        if ((candidate + "e" + (point - length)).toDouble() == value) {
            point += candidate.length - length
            digits = candidate.trimEnd('0')
            break
        }
    }
    if (point <= -6 || point > 21) {
        return digits.take(1) + (if (digits.length > 1) "." + digits.drop(1) else "") +
            "e" + (if (point > 0) "+" else "") + (point - 1)
    }
    if (point <= 0) return "0." + "0".repeat(-point) + digits
    if (point >= digits.length) return digits + "0".repeat(point - digits.length)
    return digits.take(point) + "." + digits.drop(point)
}

private fun attributes(value: Attributes): String =
    (
        value.classes.map {
            "." + identifier(it)
        } + value.records.map { "${it.name}=${escaped(it.value)}" }
    ).joinToString(" ", "{", "}")

private fun metadataValue(value: MetadataValue): String =
    when (value) {
        is MetadataValue.Scalar -> {
            "scalar(" +
                when (val scalar = value.value) {
                    MetadataScalar.Null -> "null"
                    is MetadataScalar.Bool -> "bool(${scalar.value})"
                    is MetadataScalar.Number -> "number(${escaped(scalar.value)})"
                    is MetadataScalar.Text -> "text(${escaped(scalar.value)})"
                } + ")"
        }

        is MetadataValue.List -> {
            "list([" +
                value.items.joinToString(",") { item ->
                    when (item) {
                        is MetadataListItem.Number -> "number(${escaped(item.value)})"
                        is MetadataListItem.Text -> "text(${escaped(item.value)})"
                    }
                } + "])"
        }
    }

private fun identifier(value: String): String {
    val plain =
        value.isNotEmpty() &&
            value.all { it.code in 33..126 && it.code !in listOf(34, 92, 123, 125, 91, 93, 40, 41, 61) }
    return if (plain) value else escaped(value)
}

private fun dimensions(value: Dimensions?): String =
    value?.let { "(width=${it.width},height=${it.height ?: "null"})" } ?: "null"
